File: mobile/src/main/java/guepardoapps/whosbirthday/activities/ActivityMain.java
package guepardoapps.whosbirthday.activities;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;

import guepardoapps.whosbirthday.R;
import guepardoapps.whosbirthday.common.*;
import guepardoapps.whosbirthday.controller.DatabaseController;
import guepardoapps.whosbirthday.controller.NavigationController;
import guepardoapps.whosbirthday.customadapter.BirthdayListAdapter;
import guepardoapps.whosbirthday.tools.Logger;

public class ActivityMain extends Activity {
    private static final String TAG = ActivityMain.class.getSimpleName();
    private Logger _logger;

    private boolean _created;

    private Context _context;
    private DatabaseController _databaseController;
    private NavigationController _navigationController;

    private ProgressBar _progressBar;
    private ListView _listView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.side_main);

        _logger = new Logger(TAG, Enables.LOGGING);
        _logger.Debug("onCreate");

        _context = this;

        _databaseController = DatabaseController.getInstance();
        _databaseController.Initialize(_context);

        _navigationController = new NavigationController(_context);

        _listView = findViewById(R.id.listView);

        Button btnAdd = findViewById(R.id.goToAddView);
        btnAdd.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _logger.Debug("_btnAdd onClick");
                _navigationController.NavigateTo(ActivityAdd.class, false);
            }
        });

        Button btnImpressum = findViewById(R.id.btnImpressum);
        btnImpressum.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _logger.Debug("_btnImpressum onClick");
                _navigationController.NavigateTo(ActivityImpressum.class, false);
            }
        });

        _progressBar = findViewById(R.id.progressBar);

        _created = true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        _logger.Debug("onPause");
        _databaseController.Dispose();
    }

    @Override
    protected void onResume() {
        super.onResume();
        _logger.Debug("onResume");

        _databaseController.Initialize(_context);

        if (_created) {
            _listView.setAdapter(new BirthdayListAdapter(_context, _databaseController.GetBirthdays()));
            _progressBar.setVisibility(View.GONE);
            _listView.setVisibility(View.VISIBLE);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        _logger.Debug("onDestroy");
        _databaseController.Dispose();
    }
}

File: mobile/src/main/java/guepardoapps/whosbirthday/activities/ActivityMain.kt
package guepardoapps.whosbirthday.activities

import android.app.Activity
import android.os.Bundle
import android.view.View
import guepardoapps.whosbirthday.R
import guepardoapps.whosbirthday.controller.NavigationController
import guepardoapps.whosbirthday.customadapter.BirthdayListAdapter
import guepardoapps.whosbirthday.database.birthday.DbBirthday
import guepardoapps.whosbirthday.enum.DbBirthdayAction
import guepardoapps.whosbirthday.model.DbBirthdayActionPublishSubject
import guepardoapps.whosbirthday.utils.Logger
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import it.sephiroth.android.library.bottomnavigation.BottomNavigation
import kotlinx.android.synthetic.main.side_main.*

@ExperimentalUnsignedTypes
class ActivityMain : Activity(), BottomNavigation.OnMenuItemSelectionListener {
    private val tag: String = ActivityMain::class.java.simpleName

    private var subscription: Disposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.side_main)
        bottomNavigation.setOnMenuItemClickListener(this)
        subscription = DbBirthdayActionPublishSubject.instance.publishSubject
                .subscribeOn(Schedulers.io())
                .subscribe(
                        { dbBirthdayAction ->
                            when (dbBirthdayAction) {
                                DbBirthdayAction.Add -> listView.adapter = BirthdayListAdapter(this, DbBirthday(this).get().toTypedArray())
                                DbBirthdayAction.Update -> listView.adapter = BirthdayListAdapter(this, DbBirthday(this).get().toTypedArray())
                                DbBirthdayAction.Delete -> listView.adapter = BirthdayListAdapter(this, DbBirthday(this).get().toTypedArray())
                                else -> Logger.instance.verbose(tag, "No action needed for dbBirthdayAction $dbBirthdayAction")
                            }
                        },
                        { }
                )
    }

    override fun onResume() {
        super.onResume()
        if (subscription != null) {
            listView.adapter = BirthdayListAdapter(this, DbBirthday(this).get().toTypedArray())
            listView.visibility = View.VISIBLE
            progressBar.visibility = View.GONE
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        subscription?.dispose()
        subscription = null
    }

    override fun onMenuItemSelect(itemId: Int, position: Int, fromUser: Boolean) {
        performMenuAction(itemId)
    }

    override fun onMenuItemReselect(itemId: Int, position: Int, fromUser: Boolean) {
        performMenuAction(itemId)
    }

    private fun performMenuAction(itemId: Int) {
        when (itemId) {
            R.id.item_add -> NavigationController(this).navigate(ActivityEdit::class.java, false)
            R.id.item_settings -> NavigationController(this).navigate(ActivitySettings::class.java, false)
            R.id.item_about -> NavigationController(this).navigate(ActivityAbout::class.java, false)
            else -> Logger.instance.error(tag, "Found no menu entry with id $itemId")
        }
    }
}